grpc/shared/rollupdata_handling.go

```diff
@@ -0,0 +1,261 @@
+	defer allocationUnbundlingTimer.UpdateSince(unbundlingStart)
+
+	processedTxs := types.Transactions{}
+	bid := &auctionv1alpha1.Bid{}
+
+	unprocessedBid := allocation.GetBid()
+
+	err := anypb.UnmarshalTo(unprocessedBid, bid, proto2.UnmarshalOptions{
+		Merge:        false,
+		AllowPartial: false,
+	})
+	if err != nil {
+		return nil, WrapError(err, "failed to unmarshal bid")
+	}
+
+	log.Debug("Found a potential allocation in the rollup data. Checking if it is valid.", "prevBlockHash", common.BytesToHash(prevBlockHash).String(), "auctioneerBech32Address", auctioneerBech32Address)
+
+	if !bytes.Equal(bid.GetRollupParentBlockHash(), prevBlockHash) {
+		allocationsWithInvalidPrevBlockHash.Inc(1)
+		return nil, errors.New("prev block hash in allocation does not match the previous block hash")
+	}
+
+	publicKey := ed25519.PublicKey(allocation.GetPublicKey())
+	bech32Address, err := EncodeFromPublicKey(addressPrefix, publicKey)
+	if err != nil {
+		return nil, WrapError(err, fmt.Sprintf("failed to encode public key to bech32m address: %s", publicKey))
+	}
+
+	if auctioneerBech32Address != bech32Address {
+		allocationsWithInvalidPubKey.Inc(1)
+		return nil, fmt.Errorf("address in allocation does not match auctioneer address. expected: %s, got: %s", auctioneerBech32Address, bech32Address)
+	}
+
+	message, err := proto.Marshal(bid)
+	if err != nil {
+		return nil, WrapError(err, "failed to marshal allocation to verify signature")
+	}
+
+	signature := allocation.GetSignature()
+	if !ed25519.Verify(publicKey, message, signature) {
+		allocationsWithInvalidSignature.Inc(1)
+		return nil, fmt.Errorf("signature in allocation does not match the public key")
+	}
+
+	log.Debug("Allocation is valid. Unmarshalling the transactions in the bid.")
+	// unmarshall the transactions in the bid
+	for _, allocationTx := range bid.GetTransactions() {
+		ethtx := new(types.Transaction)
+		err := ethtx.UnmarshalBinary(allocationTx)
+		if err != nil {
+			return nil, WrapError(err, "failed to unmarshall allocation transaction")
+		}
+		processedTxs = append(processedTxs, ethtx)
+	}
+
+	successfulUnbundledAllocations.Inc(1)
+
+	return processedTxs, nil
+
+}
+
+// `UnbundleRollupDataTransactions` takes in a list of rollup data transactions and returns the corresponding
+// list of Ethereum transactions.
+// If it finds any `Allocation` type, it validates it and places the txs in the `Allocation` at the top of block.
+// Note that `UnbundleRollupDataTransactions` does not return any error on an invalid `RollupData`. If we find any invalid
+// `RollupData` we log the error and continue processing the rest of the transactions. We do not want to break control flow
+// for an invalid transaction as we do not want to interrupt block production.
+// TODO - This function has too many arguments. We should consider refactoring it.
+func UnbundleRollupDataTransactions(txs []*sequencerblockv1.RollupData, height uint64, bridgeAddresses map[string]*params.AstriaBridgeAddressConfig,
```

Comment:
don't need to export

Reply:
need to export cause the execution package and the optimistic package use it